### ISSUE TITLE
Consumer stuck in infinite loop (?)

### ISSUE THREAD

**Issue by hemidactylus** [Parsely/pykafka#502]
Hello,
I am having some trouble with what I believe to be a rather simple pykafka use case.

I have set up a loop that goes on retrieving messages from a topic (one producer, one partition, no replication, only one consumer having exclusive access to the server), but sometimes it sort of gets stuck in a loop.
I have wrapped the usage of pykafka within a class of my own that, among other things, prints the offset of the retrieved message along with the first characters of its body and from time to time the offset coming out are arranged in an endlessly repeating sequence like:
 ...
 99
100
101
102
103
104
101 ( x )
102
103
104
101 ( x )
102
103
104
102 ( xx )
103
104
101 ( xxx )
102
103
104
 ...

(I omit the text of the messages, they correspond 1:1 to the offsets).
(In real cases the loop has an extension up to few hundreds of messages. My current offsets are in the half-millions range).
This happens regardless of whether I manually invoke commit_offsets() or set auto commit with a few seconds interval; the exact offset where the sequence restarts is not always the same (e.g. above it restarts from 101 in places marked x and xxx but from 102 in xx ).

I believe this is a very strange behaviour since it forces me to manually store the highest offset seen so far and check against it: whenever I detect this infinite looping behaviour, since commits are not effective, I have to stop(), wait some seconds, and start() again the simple_consumer.

I noticed this tends to happen (but perhaps not exclusively) when the consumer has reached the current end of the topic: however, if new messages are produced to it when the consumer is looping, this is not sufficient to make it exit towards the fresh messages.

Here are the kafka_server/pykafka/python versions I tested this on:
* kafka server version 0.9.0.1
* Python 2.7.10 (default, Oct 14 2015, 16:09:02)  [GCC 5.2.1 20151010] on linux2
* pykafka==2.2.1

I provide below what I deem the relevant parts of the class that provides an interface between pykafka and my other code. I'd be glad if someone can point me to the right direction -- it might well be that this is indeed the desired behaviour, in which case I may just need to rethink some parts of my code.

CLASS CODE:
```python
from pykafka import KafkaClient
kserver_default_string='RE.DA.CT.ED:port'

socket_timeout_ms=30000
offsets_channel_socket_timeout_ms=10000

class kafka_agent:

    def __init__(self, topic, kserver=ks_default_string, consumer_timeout_ms=350):
        self.topic=topic
        self.kserver=kserver
        self.kclient=KafkaClient(hosts=kserver,socket_timeout_ms=socket_timeout_ms,offsets_channel_socket_timeout_ms=offsets_channel_socket_timeout_ms)
        self.ktopic=self.kclient.topics[topic]
        self.kconsumer=self.ktopic.get_simple_consumer(consumer_group='cgroup_'+topic,
            consumer_timeout_ms=consumer_timeout_ms,auto_commit_enable=True, auto_commit_interval_ms=10000)

    def retrieve(self, block=True):
        if not self.kconsumer._running:
            self.kconsumer.start()
        tmsg=self.kconsumer.consume(block=block)
        if tmsg is not None:
            tstring=tmsg.value
            toff=tmsg.offset
            return tstring
        else:
            return None

    def commit_last_retrieve(self):
        self.kconsumer.commit_offsets()

    def isrunning(self):
        return self.kconsumer._running

    # those are for usage by context managers
    def __exit__(self, exc_type, exc_value, traceback):
        self.kconsumer.stop()

    def __enter__(self):
        self.kconsumer.start()
```

USER CODE: something along the lines of:
```python

with kafka_agent('TOPIC_NAME') as kagent:
    while True:
        obj=kagent.retrieve()
        if obj is not None:
            # do stuff with obj
        else:
            # None was retrieved
        kagent.commit_last_retrieve()
```

===

Thanks a lot!

**Comment by emmett9001** [comment]
Hi @hemidactylus. I don't see anything in your code that I'd expect to cause this looping behavior. If there's some place in your code where you're destroying and recreating the `SimpleConsumer` instance, it's definitely possible for this issue to occur. Maybe you can figure out if you've got any logic in your code that would cause [these checks](https://github.com/Parsely/pykafka/blob/412466e2152d061339891fb24ef4ec5709e89782/pykafka/simpleconsumer.py#L848) not to function as intended.

**Comment by hemidactylus** [comment]
Hallo @emmett9001 and thanks for the quick reply! I do not think the consumer is reinstantiated at a fast pace, however I noticed that the code as it is issues a lot of commit_offsets() calls, up to several per second over the internet to the server. I suspect that may somehow lead to more than one commit traveling at the same time to the server and piling up in an unclean way, so that the local client code may end up retaining offset information that does not correspond to the latest commit I sent.

Considering I set auto_commit_enable=True, I am now testing the same code as above with no commits by hand and the issue seems not to show up anymore. I'll let you know if I find something that might be related to wrongdoings taking place outside of my wrapper class.

Thanks again and have a nice day!
Stefano
